src/ws/session.c

```diff
@@ -515,12 +515,13 @@ open_session (pam_handle_t *pamh)
       return res;
     }
 
-  /* Yes, buf "leaks" */
-  buf = malloc (sizeof (struct passwd) + 8192);
-  if (buf == NULL)
+  if (!pwd_buf)
+    pwd_buf = malloc (sizeof (struct passwd) + 8192);
+
+  if (pwd_buf == NULL)
     res = ENOMEM;
   else
-    res = getpwnam_r (name, buf, (char *)(buf + 1), 8192, &pwd);
+    res = getpwnam_r (name, pwd_buf, (char *)(pwd_buf + 1), 8192, &pwd);
```

Comment:
I had to read this pointer arithmetic wizardry three times to confirm it looks correct. IMHO it wouldn't kill us to drop this, the allocations, and just have a `static struct passwd pwd;` and `static char pwd_buf[8192];` at the top of the file, and use `sizeof` :-)